sdks/python/apache_beam/typehints/sharded_key_type.py

```diff
@@ -0,0 +1,69 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pytype: skip-file
+
+from __future__ import absolute_import
+
+from apache_beam.coders import typecoders
+from apache_beam.coders.coders import ShardedKeyCoder
+from apache_beam.typehints import typehints
+from apache_beam.utils.sharded_key import ShardedKey
+
+
+class ShardedKeyTypeConstraint(typehints.TypeConstraint):
+  def __init__(self, key_type):
+    self.key_type = typehints.normalize(key_type)
+
+  def _inner_types(self):
+    yield self.key_type
+
+  def _consistent_with_check_(self, sub):
+    return (
+        isinstance(sub, self.__class__) and
+        typehints.is_consistent_with(sub.key_type, self.key_type))
+
+  def type_check(self, instance):
```

Comment:
The 2 exceptions raised here should have test coverage.

Reply:
Done. Added test coverage and some minor fixes in the latest commit.